braid/braid.jl/wrapper_functions.jl

```diff
@@ -0,0 +1,310 @@
+#=BHEADER**********************************************************************
+ * Copyright (c) 2013, Lawrence Livermore National Security, LLC. 
+ * Produced at the Lawrence Livermore National Laboratory.
+ * 
+ * This file is part of XBraid. For support, post issues to the XBraid Github page.
+ * 
+ * This program is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License (as published by the Free Software
+ * Foundation) version 2.1 dated February 1999.
+ * 
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY
+ * WARRANTY; without even the IMPLIED WARRANTY OF MERCHANTABILITY or FITNESS FOR A
+ * PARTICULAR PURPOSE. See the terms and conditions of the GNU General Public
+ * License for more details.
+ * 
+ * You should have received a copy of the GNU Lesser General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc., 59
+ * Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ *
+ ***********************************************************************EHEADER=#
+
+"""
+Displays a caught error message, including stacktrace, without throwing
+"""
+function stacktrace_warn(msg::String, err)
+    err_msg = sprint(showerror, err)
+    trace = sprint((io,v) -> show(io, "text/plain", v), stacktrace(catch_backtrace()))
+    @warn "$(msg):\n$(err_msg)\n$(trace)"
+end
+
+"""
+Where'd all my data go?
+This extends Base to include a buffer which throws away the data written to it
+(useful for measuring the serialized size of an object)
+"""
+mutable struct BlackHoleBuffer <: IO
+    ptr::Int
+end
+BlackHoleBuffer() = BlackHoleBuffer(0)
+
+function Base.read(from::BlackHoleBuffer, T::Type{UInt8})
+    throw(ArgumentError("BlackHoleBuffer is not readable)"))
+end
+function Base.write(to::BlackHoleBuffer, x::UInt8)
+    to.ptr += 1
+    return 1
+end
+function Base.write(to::BlackHoleBuffer, x::Array{T}) where T
+    to.ptr += sizeof(x)
+    return sizeof(x)
+end
+
+# these are internal functions which directly interface with XBraid
+
+function _jl_step!(_app::Ptr{Cvoid},
+                   _ustop::Ptr{Cvoid},
+                   _fstop::Ptr{Cvoid},
+                   _u::Ptr{Cvoid},
+                   _status::Ptr{Cvoid})::Cint
+    # println("step")
+    app = unsafe_pointer_to_objref(_app)::BraidApp
+    u = unsafe_pointer_to_objref(_u)::BraidVector
+    ustop = unsafe_pointer_to_objref(_ustop)::BraidVector
+    status = StepStatus(_status)
+
+    tstart, tstop = status_GetTstartTstop(status)
+    delta_rank = status_GetDeltaRank(status)
+
+    # call the user's function
+    try
+        # residual option
+        if _fstop !== C_NULL
```

Comment:
Julia uses "multiple dispatch," i.e. function overloading, rather than null values to handle function arguments, so the user has to make sure their step function takes the right number and types of arguments.

Reply:
Makes sense to me